flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java

```diff
@@ -125,4 +160,82 @@ public void validateColumnListParams(
         // this makes it possible to ignore them in the validator and fall back to regular row types
         // see also SqlFunction#deriveType
     }
+
+    protected void registerNamespace(
+            @Nullable SqlValidatorScope usingScope,
+            @Nullable String alias,
+            SqlValidatorNamespace ns,
+            boolean forceNullable) {
+
+        // apply snapshot to SqlValidatorNameSpace
+        // Time travel only supports constant expressions, so we need to investigate scenarios
+        // where the period of Snapshot is a SqlIdentifier.
+        if (usingScope != null
+                && ns instanceof IdentifierNamespace
+                && ns.getEnclosingNode() instanceof SqlSnapshot
+                && !(((SqlSnapshot) ns.getEnclosingNode()).getPeriod() instanceof SqlIdentifier)) {
+            SqlSnapshot sqlSnapshot = (SqlSnapshot) ns.getEnclosingNode();
+            SqlNode periodNode = sqlSnapshot.getPeriod();
+            SqlToRelConverter sqlToRelConverter = this.createSqlToRelConverter();
+            RexNode rexNode = sqlToRelConverter.convertExpression(periodNode);
+            RexNode simplifiedRexNode =
+                    FlinkRexUtil.simplify(
+                            sqlToRelConverter.getRexBuilder(),
+                            rexNode,
+                            relOptCluster.getPlanner().getExecutor());
+            List<RexNode> reducedNodes = new ArrayList<>();
+            relOptCluster
+                    .getPlanner()
+                    .getExecutor()
+                    .reduce(
+                            relOptCluster.getRexBuilder(),
+                            Collections.singletonList(simplifiedRexNode),
+                            reducedNodes);
+            // check whether period is the unsupported expression
+            if (!(reducedNodes.get(0) instanceof RexLiteral)) {
+                throw new UnsupportedOperationException(
+                        String.format("Unsupported time travel period: %s", periodNode));
+            }
+
+            RexLiteral rexLiteral = (RexLiteral) (reducedNodes).get(0);
+            sqlSnapshot.setOperand(
+                    1,
+                    SqlLiteral.createTimestamp(
+                            rexLiteral.getValueAs(TimestampString.class),
+                            rexLiteral.getType().getPrecision(),
+                            sqlSnapshot.getPeriod().getParserPosition()));
+
+            TimestampString timestampString = rexLiteral.getValueAs(TimestampString.class);
+            TableConfig tableConfig = ShortcutUtils.unwrapContext(relOptCluster).getTableConfig();
+            ZoneId zoneId = tableConfig.getLocalTimeZone();
+
+            long timeTravelTimestamp =
+                    TimestampData.fromEpochMillis(timestampString.getMillisSinceEpoch())
```

Comment:
dito:
Also check timestampString is not null to reduce the warning  `Method invocation 'getMillisSinceEpoch' may produce 'NullPointerException' ` from IDE